dotnet-desktop-guide/net/wpf/documents/how-to-print-xps-files.md

```diff
@@ -1,28 +1,59 @@
   - "csharp"
   - "vb"
 helpviewer_keywords:
-  - 
+  - "programmatically print xps files [WPF]"
 ---
 
 # How to print an XPS file (WPF .NET)
 
-<!--
+Sometimes you'll want to add a new print job to the print queue without opening a print dialog box. You can use one of the [PrintQueue.AddJob](/dotnet/api/system.printing.printqueue.addjob) methods to do that. Here's how.
 
-Source material for this article:
-- https://docs.microsoft.com/en-us/dotnet/desktop/wpf/advanced/how-to-programmatically-print-xps-files?view=netframeworkdesktop-4.8
+[!INCLUDE [desktop guide under construction](../../includes/desktop-guide-preview-note.md)]
 
-Notes:
-- None
+In the following example, we use the <xref:System.Printing.PrintQueue.AddJob%28System.String%2CSystem.String%2CSystem.Boolean%29> method, one of the several overloads of `AddJob`, to:
 
--->
+- Add a new print job for an XML Paper Specification (XPS) document into the default print queue.
+- Name the new job.
+- Specify whether or not the XPS document should be validated (by using the `fastCopy` parameter).
 
-!!INTRO
+When using the `AddJob(String, String, Boolean)` method, the value of the `fastCopy` parameter is a key consideration:
 
-[!INCLUDE [desktop guide under construction](../../includes/desktop-guide-preview-note.md)]
+- If you set the `fastCopy` parameter to `true`, then XPS validation is skipped and the print job will spool quickly without page-by-page progress feedback.
+- If you set the `fastCopy` parameter to `false`, then the thread that calls the `AddJob` method must have a [single-threaded apartment state](/dotnet/api/system.threading.apartmentstate), otherwise an exception will be thrown. For more about this, see the **Remarks** section for <xref:System.Printing.PrintQueue.AddJob%28System.String%2CSystem.String%2CSystem.Boolean%29>.
+
+## Add new print jobs to the queue
+
+This example adds one or more XPS documents to the default queue. The code will:
+
+1. Use [Task.Run](/dotnet/api/system.threading.tasks.task.run) to avoid blocking the UI thread&mdash;since there is no async version of `AddJob`.
+1. If the `fastCopy` parameter value is `false`, run <xref:System.Printing.PrintQueue.AddJob%28System.String%2CSystem.String%2CSystem.Boolean%29> on a thread with [single-threaded apartment state](/dotnet/api/system.threading.apartmentstate).
+1. Get a reference to the default <xref:System.Printing.PrintQueue> of the <xref:System.Printing.LocalPrintServer>.
+1. Call `AddJob(String, String, Boolean)` on the print queue reference, passing in a job name, an XPS document path, and the `fastCopy` parameter.
+
+If the queue is not paused and the printer is working, then a print job will automatically begin printing when it reaches the top of the print queue.
+
+> [!TIP]
+> To avoid the **Save Output File As** dialog when adding a print job to the default queue, ensure that your default printer is not **Microsoft XPS Document Writer**, **Microsoft Print to PDF**, or other print-to-file option.
```

Comment:
```suggestion
> To avoid the **Save Output File As** dialog when adding a print job to the default queue, ensure that your default printer is not **Microsoft XPS Document Writer**, **Microsoft Print to PDF**, or other print-to-file options.
```

Reply:
Done.